examples/captive_portal.toit

```diff
@@ -2,41 +2,134 @@
+// back on this port when testing on desktop Toit.
+HOST_PORT ::= 8080
+
+// Customize this function to insert dynamic content in files using the
+// {{variable}} syntax.
+look_up_variable variable/string -> string:
+  if variable == "time":
+    return Time.now.local.stringify
+
+  // Default - just color it red to make it easy for the developer to spot.
+  return "<font color=red>{{$variable}}</font>"
+
+TEMPORARY_REDIRECTS ::= {
+  "generate_204": "/",    // Used by Android captive portal detection.
+  "gen_204": "/",         // Used by Android captive portal detection.
+  // Add more redirects in order to create an alias for a file.
+}
+
+// If you add files to the website directory you may need to add any missing
+// file extensions here.  Use only lower case.
+MIME_TYPES ::= {
+  "txt":  "text/plain",
+  "html": "text/html",
+  "css":  "text/css",
+  "png":  "image/png",
+  "pdf":  "application/pdf",
+  "jpg":  "image/jpeg",
+  "jpeg": "image/jpeg",
+  "ico":  "image/png",
+  "svg":  "image/svg+xml",
+}
+
 main:
-  network := wifi.establish --ssid="mywifi" --password="12345678"
+  network := null
+  on_device := false
+  exception := catch --trace:
+    network = wifi.establish --ssid=CAPTIVE_PORTAL_SSID --password=CAPTIVE_PORTAL_PASSWORD
+    on_device = true
+  if exception:
+    network = net.open
 
-  task:: run_dns network
+  if on_device:
+    task:: run_dns network
 
-  server_socket := network.tcp_listen 80
-  while true:
-    socket := server_socket.accept
-    if not socket: continue
-    // The worlds worst HTTP server - you certainly want to put something
-    // better here.
-    print "Got a connection attempt on port 80 from $socket.peer_address"
-    reader := BufferedReader socket
-    while line := reader.read_line:
-      if line.ends_with "\r": line = line[..line.size - 1]
-      print "HTTP header: $line"
-      if line == "": break
-    try:
-      socket.write "HTTP/1.1 200 OK\r\n"
-      socket.write "text/plain\r\n"
-      socket.write "Content-Type: text/plain\r\n"
-      socket.write "\r\n"
-      socket.write "Welcome to Toit! $(Time.now)\r\n"
-    finally:
-      socket.close
-  server_socket.close
-  network.close
+  port := on_device ? 80 : HOST_PORT
+  server := http.Server
+  print "Listening on http://localhost:$port/"
+
+  server.listen network port:: | request writer |
+    handle request writer
+
+mime_type path/string -> string:
+  suffix := path
+  index := path.index_of --last "."
+  if index != -1:
+    suffix = path[index + 1..]
+  return MIME_TYPES.get suffix --if_absent=:
+    suffix = to_lower_case suffix
+    return MIME_TYPES.get suffix --if_absent=:
+      throw "Unknown MIME type for $path"
+
+to_lower_case in/string -> string:
+  in.do:
+    if 'A' <= it <= 'Z':
+      ba := in.to_byte_array
+      ba.size.repeat:
+        if 'A' <= ba[it] <= 'Z':
+          ba[it] |= 0x20
+      return ba.to_string
+  return in
+
+substitute_variables result/string -> string:
```

Comment:
This feels almost useful in general. Maybe something to consider (with a block) for the core libraries?